docs: unify installation (#1125)
ivadomed/ivadomed: docs/source/installation.rst

            If you are using `Compute Canada <https://www.computecanada.ca/>`_, you can load modules as `mentioned here <https://intranet.neuro.polymtl.ca/computing-resources/compute-canada#modules>`_ and `also here <https://docs.computecanada.ca/wiki/Utiliser_des_modules/en#Loading_modules_automatically>`_.

Step 2: Install ``ivadomed`` with GPU or CPU Support
++++++++++++++++++++++++++++++++++++++++++++++++++++

    .. tabs::
        
        .. tab:: NVIDIA GPU Support

            ``ivadomed`` requires a minimum PyTorch version of either
            1.8.1 or 1.8.2 which supports CUDA 10.2 and CUDA 11.1 builds
            by default. 
            
            Ampere-based GPUs (with a Compute Capability of 8.x) natively
            support CUDA>=11.1. Although CUDA 11.1 is backward compatible

>             but CUDA 10.2 is preferred for older GPUs.

Author: Interesting updates #1141 & the upstream torch thread, at least for torch ~=1.8.1

>             Ampere-based GPUs (with a Compute Capability of 8.x) natively
>             support CUDA>=11.1. Although CUDA 11.1 is backward compatible
>             but CUDA 10.2 is preferred for older GPUs.

Author: > My understanding is that it's not a question of the GPUs supporting CUDA, it's if CUDA supports the GPUs,

Agreed... looks like it didn't come through as thought, but your suggestion looks apt.

> the reason for preferring 10.2 is simply that it's the stable branch at the moment (and is for example [more optimized](https://github.com/ivadomed/ivadomed/issues/1141))

I'd guess other CUDA versions (11.1, etc) are stable as well, but it's just that this jump didn't propagate as expected until torch1.11 for older hardware. https://github.com/pytorch/pytorch/issues/47908#issuecomment-1148387263 So, skipping the reason should be good... What about?

```suggestion
            Ampere-based GPUs (with a Compute Capability of 8.x) only work with
            CUDA>=11.1. Although CUDA 11.1 is backward compatible with older hardware,
            CUDA 10.2 is preferred if available.
```

            CUDA 10.2 and CUDA 11.1 require an NVIDIA driver version >=440.33 
            and >=450 respectively as indicated `here <https://docs.nvidia.com/deploy/cuda-compatibility/index.html#minor-version-compatibility>`__.
            Please make sure to upgrade to the minimum NVIDIA driver version 
            requirements for the respective CUDA builds. To verify the NVIDIA
            driver version, just run the command `nvidia-smi` and you'll find 

>             your current driver version on the top.

Reviewer: There's also the `pip.conf` solution: instead of "please run this command" it could be "please create this file before installation"

```
# /etc/pip.conf or ~/.config/pip/pip.conf or C:\ProgramData\pip\pip.ini or %APPDATA%\pip\pip.ini
[install]
extra-index-url =
   https://download.pytorch.org/whl/cu111
   https://download.pytorch.org/whl/cu113
   https://download.pytorch.org/whl/cu115
```

That's essentially instructing people to set a global variable on their OS, but I think this is one of the rare cases where that's the right thing to do: a system has fixed hardware (usually), so every pretty much every package they use with their GPUs will need the same patch.

>             your current driver version on the top.

Author: > I also wonder if even mentioning "10.2" will mislead people. 

Maybe or maybe not

> It's not really CUDA 10.2 that's important, but rather just using the default `torch` that's on PyPI, because that's the best-tested (and evidently most optimized) version of `torch` out there. Similarly the "minimum version" of torch is encoded in the .whl, so that's also something users shouldn't need to think about. So maybe:

Agreed

> (I replaced `nvidia-smi` with `cat` because `nvidia-smi` isn't installed with the drivers; on Ubuntu it comes [with `nvidia-cuda-toolkit`](https://github.com/neuropoly/computers/blob/99c8bf5c24365f7eb1f71d252011f605ead425c7/ansible/roles/neuropoly.gpu/tasks/main.yml#L19) but not with `nvidia-driver-XXX`; 

Thank you for adding this... I was sceptical about this as well

> I suppose for someone who was trying to get things working on Windows, this is a step back, since there maybe `nvidia-smi` does come? But on Windows someone can probably look in Control Panel to find the driver version?)
Yuss yuss

> `ivadomed-test-gpus` is a new suggestion, suggesting we make sure people can actually. We both know it could be `python -c 'import torch; print(torch.cuda.is_available())`, but maybe we should wrap it in a small helper script to prove to people that ivadomed is actually talking to the GPUs? Notice that except for "download.pytorch.org" we don't need to mention torch at all -- it's just a dependency, it's not what people are looking to learn about from ivadomed so the more we avoid opening the rabbit hole the better I think. But if we did ship such a script we'd have to maitnain it; so is there an obvious command that already exists in ivadomed that reports whether it's using GPUs or not that people could run on some small test data, just to verify everything?
I agree... I'm thinking something along the lines what ADS has ([Quick Test](https://axondeepseg.readthedocs.io/en/latest/documentation.html#quick-test)). It should come after the installation step... let's skip this for now

>             your current driver version on the top.

Author: > There's also the `pip.conf` solution: instead of "please run this command" it could be "please create this file before installation"
> 
> ```
> # /etc/pip.conf or ~/.config/pip/pip.conf or C:\ProgramData\pip\pip.ini or %APPDATA%\pip\pip.ini
> [install]
> extra-index-url =
>    https://download.pytorch.org/whl/cu111
>    https://download.pytorch.org/whl/cu113
>    https://download.pytorch.org/whl/cu115
> ```
> 
> That's essentially instructing people to set a global variable on their OS, but I think this is one of the rare cases where that's the right thing to do: a system has fixed hardware (usually), so every pretty much every package they use with their GPUs will need the same patch.

I'm a little hesitant towards this solution bcz let's say someone follows these instructions and later on forgets this setting, and perhaps the torch installation evolves, then it'd be harder for them to debug.

                            ::
                                
                                pip install ivadomed --extra-index-url https://download.pytorch.org/whl/cu102

                            and, with CUDA 11.1:

                            ::

                                pip install ivadomed --extra-index-url https://download.pytorch.org/whl/cu111

                .. tab:: Source Installation

                    Bleeding-edge developments are available on the master branch of the project
                    on Github. To install ``ivadomed`` from source

                    .. tabs::

                        .. tab:: Linux

                            with CUDA 10.2:
                            
                            ::
                   
                                git clone https://github.com/ivadomed/ivadomed.git

                                cd ivadomed

                                pip install -e .

                            and, with CUDA 11.1:

                            ::

                                git clone https://github.com/ivadomed/ivadomed.git

                                cd ivadomed

>                                 pip install -e . --extra-index-url https://download.pytorch.org/whl/cu111

Reviewer 2: Good point. I like this approach of keeping things simple and ask others to do extra work. Why we keep things complicated and hard to update when we can ask the non-default path users to be doing more work right?

                        .. tab:: Windows 

                            with CUDA 10.2:

                            ::

                                git clone https://github.com/ivadomed/ivadomed.git

                                cd ivadomed

>                                 pip install -e . --extra-index-url https://download.pytorch.org/whl/cu102

Reviewer: My suggestion above drops Windows support entirely, because @jcohenadad [said we should](https://github.com/neuropoly/computers/issues/202#issuecomment-1139710695), but it would be good if the trick to getting GPUs working on Windows was saved _somewhere_.

I'm not sure where though. Maybe it could go in CONTRIBUTING.md along with the rest of this section, where we aren't as constrained by having to write for a wider audience and can dump reams of technical details with less care?

Or maybe we could tuck it into the GPU section in a `tabs:::` afterall and just ignore @jcohenadad's ask -- I wrote in my suggestion "Not officially supported nor tested", but at the moment GPU support on Linux isn't tested either, except by our own research, so maybe it's not so much of a stretch to just tuck in the Windows case too and get around to testing it [mañana](https://www.wordnik.com/words/ma%C3%B1ana).

Reviewer 2: GPU testing will resume outside of the existing setup for sanity check reasons. It will be a completely independent and private process though as our existing GitHub workflow is not flexible enough to accommodate it and cannot securely make it integrate with GitHub in its current implementation.  I will open bugs on issue found manually for now.

                            with CUDA 10.2:

                            ::

                                git clone https://github.com/ivadomed/ivadomed.git

                                cd ivadomed

                                pip install -e .[dev]

                            with CUDA 11.1:

                            ::

                                git clone https://github.com/ivadomed/ivadomed.git

                                cd ivadomed

                                pip install -e .[dev] --extra-index-url https://download.pytorch.org/whl/cu111

                        .. tab:: Windows

                            with CUDA 10.2:

                            ::

                                git clone https://github.com/ivadomed/ivadomed.git

                                cd ivadomed

                                pip install -e .[dev] --extra-index-url https://download.pytorch.org/whl/cu102

                        
                            with CUDA 11.1:

                            ::

                                git clone https://github.com/ivadomed/ivadomed.git

                                cd ivadomed

>                                 pip install -e .[dev] --extra-index-url https://download.pytorch.org/whl/cu111

Author: Absolutely... At that point, I was trying to consolidate everything.

        .. tab:: CPU Support

            .. tabs:: 
                
                .. tab:: Package Installation (Recommended)

                    .. tabs::

                        .. tab:: Linux

                            ::

>                                 pip install ivadomed --extra-index-url https://download.pytorch.org/whl/cpu

Reviewer: I'm personally invested in this one line, because we [use it in SCT](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3790), but maybe it's better to drop it. The torch people think linux people should waste their space, so let's just roll with that?

Alternately, this could fit in a `tabs:::` above, next to the GPU section, maybe something like:

> Installation | CPU-based Installation | GPU-based Installation (Windows) | Ampere GPU-based Installation

So the 80% case is the first tab, which wastes some Linux users' space who haven't read closely but that's not the end of the world, the 15% case is the people who read the CPU instructions, another say ~5% of people are bold explorers using GPUs on Windows, and the <1% case is the people rich enough to have Ampere GPUs.

And the only difference between each tab is

a. Instructing people how to decide if that case applies to them
b. Giving different `extra-index-url` lines for their `pip.conf`

Author: Flattening the tab hierarchy would result in repetition of some GPU instructions.

                                cd ivadomed

                                pip install -e .

                .. tab:: Contributor or Developer Installation

                    To contribute to the project, we recommend installing ``ivadomed`` from source along with additional dependencies related to building documentation and testing:

                    .. tabs::

                        .. tab:: Linux
                            
                            ::

                                git clone https://github.com/ivadomed/ivadomed.git

                                cd ivadomed

                                pip install -e .[dev] --extra-index-url https://download.pytorch.org/whl/cpu

                        .. tab:: Windows 

                            ::

                                git clone https://github.com/ivadomed/ivadomed.git

                                cd ivadomed

                                pip install -e .[dev]

                        .. tab:: Mac

                            ::

                                git clone https://github.com/ivadomed/ivadomed.git

                                cd ivadomed

>                                 pip install -e .[dev]

Reviewer: The diff makes it hard to comment on the right part here, but again, I'd put all this in CONTRIBUTING.md.

            If you are using `Compute Canada <https://www.computecanada.ca/>`_, you can load modules as `mentioned here <https://intranet.neuro.polymtl.ca/computing-resources/compute-canada#modules>`_ and `also here <https://docs.computecanada.ca/wiki/Utiliser_des_modules/en#Loading_modules_automatically>`_.

Step 2: Install ``ivadomed`` with GPU or CPU Support
++++++++++++++++++++++++++++++++++++++++++++++++++++

    .. tabs::

        .. tab:: NVIDIA GPU Support

            ``ivadomed`` requires CUDA 11.1 to execute properly. If you have an NVIDIA GPU, look up its Cuda Compute Score `here <https://developer.nvidia.com/cuda-gpus>`__, which needs to be > 3.5 to be compatible with CUDA 11.1. Also, make sure to upgrade your NVIDIA driver to at least the minimum required version as indicated `here <https://docs.nvidia.com/deploy/cuda-compatibility/index.html#minor-version-compatibility>`__.

            To verify the CUDA version, simply run ``nvcc -V`` in both Linux and Windows. The release refers to your current CUDA version.

            .. tabs::

>                 .. tab:: Package Installation (Recommended)

Reviewer 2: See below.

                    After verifying CUDA 11.1 installation, install ``ivadomed`` from `PyPI <https://pypi.org/project/ivadomed/>`__:
                    
                    ::

                        pip install --upgrade pip

                        pip install ivadomed[gpu]

>                 .. tab:: Source Installation

Reviewer 2: Could try ` .. group-tab:: Source Installation` and allow sync between the GPU and CPU page when switching.

                    Bleeding-edge developments are available on the master branch of the project
                    on Github. To install ``ivadomed`` from source:

                    ::

                        git clone https://github.com/ivadomed/ivadomed.git

                        cd ivadomed

                        pip install -e .[gpu]

>                 .. tab:: Contributor or Developer Installation

Reviewer 2: Ditto.

                    To contribute to the project, we recommend installing ``ivadomed`` from source along with additional dependencies related to building documentation and testing:

                    ::
                        
                        git clone https://github.com/ivadomed/ivadomed.git

                        cd ivadomed

                        pip install -e .[dev_gpu]

>         .. tab:: CPU Support

Reviewer 2: Ditto.

            .. tabs:: 
                

>                 .. tab:: Package Installation (Recommended)

Reviewer 2: Ditto.

                    Install ``ivadomed`` from `PyPI <https://pypi.org/project/ivadomed/>`__:
                    
                    ::

                        pip install --upgrade pip

                        pip install ivadomed[cpu]

>                 .. tab:: Source Installation

Reviewer 2: Ditto.

                    Bleeding-edge developments are available on the project's master branch
                    on Github. To install ``ivadomed`` from source:

                    ::

                        git clone https://github.com/ivadomed/ivadomed.git

                        cd ivadomed

                        pip install -e .[cpu]

>                 .. tab:: Contributor or Developer Installation

Reviewer 2: Ditto.
